### PR TITLE
Ensure conversion to `flex.int` via conversion to `np.intc`

### LINKED ISSUE #593: Flumpy (?) long vs int issue on Windows
The following error from the dxtbx conda feedstock Windows build is (I think) likely an issue with flumpy incorrectly converting numpy ints to `flex.long` rather than `flex.int` as on other platforms. This hasn't been noticed till now because [we haven't had working Windows CI for some time](https://github.com/cctbx/dxtbx/issues/396).
```
_______________________ test_get_raw_data_single_panel ________________________

nxdata_example = <HDF5 file " " (mode r+)>

    def test_get_raw_data_single_panel(nxdata_example):
        det = dxtbx.nexus.nxmx.NXdetector(nxdata_example["/entry/instrument/detector"])
        nxdata = dxtbx.nexus.nxmx.NXdata(nxdata_example["/entry/data"])
        for i in range(3):
            raw_data = dxtbx.nexus.get_raw_data(nxdata, det, i)
            assert len(raw_data) == 1
>           assert isinstance(raw_data[0], flex.int)
E           AssertionError: assert False
E            +  where False = isinstance(<scitbx_array_family_flex_ext.long object at 0x00000280E923DB40>, <class 'scitbx_array_family_flex_ext.int'>)
E            +    where <class 'scitbx_array_family_flex_ext.int'> = flex.int

dxtbx\tests\nexus\test_build_dxtbx_models.py:570: AssertionError
```

### PR DESCRIPTION
**Pull request by rjgildea**
On Linux `np.int32` is equivalent to `np.intc` and hence an `int` C-type. On Windows `np.int32` is equivalent to `np.int_` and hence a `long` C-type.

Explicitly converting to `np.intc` ensures it is represented as an `int` C-type on all platforms.

Fixes #593

### CODE REVIEW

**Comment by rjgildea** [comment]
On Windows:
```
python
Python 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 08:41:22) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> from dxtbx import flumpy
>>> a = np.array([1, 2, 3])
>>> flumpy.from_numpy(a.astype(np.int32))
<scitbx_array_family_flex_ext.long object at 0x000001C828D96840>
>>> flumpy.from_numpy(a.astype(np.int_))
<scitbx_array_family_flex_ext.long object at 0x000001C828D96B40>
>>> flumpy.from_numpy(a.astype(np.intc))
<scitbx_array_family_flex_ext.int object at 0x000001C828D96840>
```
On Linux:
```
$ dials.python
Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:35:26) [GCC 10.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> from dxtbx import flumpy
>>> a = np.array([1, 2, 3])
>>> flumpy.from_numpy(a.astype(np.int32))
<scitbx_array_family_flex_ext.int object at 0x7f0c35897290>
>>> flumpy.from_numpy(a.astype(np.int_))
<scitbx_array_family_flex_ext.long object at 0x7f0c35897400>
>>> flumpy.from_numpy(a.astype(np.intc))
<scitbx_array_family_flex_ext.int object at 0x7f0c35897290>
```
See also:
https://numpy.org/doc/stable/user/basics.types.html
https://www.ibm.com/docs/en/ibm-mq/7.5?topic=platforms-standard-data-types

**Review by graeme-winter — COMMENTED** [review]
I think I need some more information / context here: from what I can tell you are going from something which was explicit to something which is now platform defined in the name of portability.

**Comment by rjgildea** [comment]
More context is given in the linked issue https://github.com/cctbx/dxtbx/issues/593 - the current behaviour _is_ platform dependent because on Unix an `np.int32` gets converted to a `flex.int`, whereas on Windows it gets converted to a `flex.long` which downstream code can’t handle. I.e. the existing code is broken on Windows.

**Comment by rjgildea** [comment]
See e.g. [failing dials-feedstock tests on Windows](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=624755&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213):
```
2022-12-07T19:35:08.7354273Z ____________________ test_convert_to_cbf[image_15799.nxs] _____________________
2022-12-07T19:35:08.7355134Z [gw0] win32 -- Python 3.10.8 %PREFIX%\python.exe
2022-12-07T19:35:08.7355437Z 
2022-12-07T19:35:08.7355835Z dials_data = <DataFetcher: C:\Users\VssAdministrator\.cache\dials_data>
2022-12-07T19:35:08.7356339Z filename = 'image_15799.nxs'
2022-12-07T19:35:08.7356932Z tmp_path = WindowsPath('C:/Users/VssAdministrator/AppData/Local/Temp/pytest-of-VssAdministrator/pytest-0/popen-gw0/test_convert_to_cbf_image_15791')
2022-12-07T19:35:08.7357401Z 
2022-12-07T19:35:08.7357823Z     @pytest.mark.parametrize("filename", ["image_15799_master.h5", "image_15799.nxs"])
2022-12-07T19:35:08.7358824Z     def test_convert_to_cbf(dials_data, filename, tmp_path):
2022-12-07T19:35:08.7359491Z         result = procrunner.run(
2022-12-07T19:35:08.7359879Z             [
2022-12-07T19:35:08.7360305Z                 "dials.import",
2022-12-07T19:35:08.7360807Z                 dials_data("vmxi_thaumatin", pathlib=True) / filename,
2022-12-07T19:35:08.7361287Z                 "image_range=1,10",
2022-12-07T19:35:08.7361972Z             ],
2022-12-07T19:35:08.7362485Z             working_directory=tmp_path,
2022-12-07T19:35:08.7363065Z         )
2022-12-07T19:35:08.7363455Z         result.check_returncode()
2022-12-07T19:35:08.7363927Z         assert not result.stderr
2022-12-07T19:35:08.7364392Z         assert tmp_path.joinpath("imported.expt").is_file()
2022-12-07T19:35:08.7364892Z     
2022-12-07T19:35:08.7365211Z         result = procrunner.run(
2022-12-07T19:35:08.7365966Z             ["dials.convert_to_cbf", "imported.expt"], working_directory=tmp_path
2022-12-07T19:35:08.7366578Z         )
2022-12-07T19:35:08.7367380Z >       result.check_returncode()
2022-12-07T19:35:08.7367726Z 
2022-12-07T19:35:08.7368257Z dials\tests\command_line\test_convert_to_cbf.py:24: 
2022-12-07T19:35:08.7369199Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2022-12-07T19:35:08.7369529Z 
2022-12-07T19:35:08.7370187Z self = ReturnObject(args=('D:\\bld\\dials_1670437421753\\_test_env\\Scripts\\dials.convert_to_cbf.EXE', 'imported.expt'), ret... dxtbx Error: D:\\bld\\dxtbx_1670421694546\\work\\dxtbx\\src\\dxtbx\\..\\dxtbx/imageset.h(473): Unknown type long\r\n')
2022-12-07T19:35:08.7371130Z 
2022-12-07T19:35:08.7371498Z     def check_returncode(self):
2022-12-07T19:35:08.7371877Z         """Raise CalledProcessError if the exit code is non-zero."""
2022-12-07T19:35:08.7372286Z         if self.returncode:
2022-12-07T19:35:08.7372936Z >           raise CalledProcessError(self.returncode, self.args, self.stdout,
2022-12-07T19:35:08.7373403Z                                      self.stderr)
2022-12-07T19:35:08.7373971Z E           subprocess.CalledProcessError: Command '('D:\\bld\\dials_1670437421753\\_test_env\\Scripts\\dials.convert_to_cbf.EXE', 'imported.expt')' returned non-zero exit status 1.
2022-12-07T19:35:08.7374761Z ..\_test_env\lib\subprocess.py:457: CalledProcessError
2022-12-07T19:35:08.7375270Z ---------------------------- Captured stdout call -----------------------------
2022-12-07T19:35:08.7375794Z DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
2022-12-07T19:35:08.7376423Z DIALS 3.11
2022-12-07T19:35:08.7376985Z The following parameters have been modified:
2022-12-07T19:35:08.7377883Z input {
2022-12-07T19:35:08.7378635Z   experiments = <image files>
2022-12-07T19:35:08.7379260Z }
2022-12-07T19:35:08.7379826Z geometry {
2022-12-07T19:35:08.7380455Z   scan {
2022-12-07T19:35:08.7380988Z     image_range = 1 10
2022-12-07T19:35:08.7381490Z   }
2022-12-07T19:35:08.7382012Z }
2022-12-07T19:35:08.7383562Z Applying input geometry in the following order:
2022-12-07T19:35:08.7384155Z   1. Manual geometry
2022-12-07T19:35:08.7395986Z --------------------------------------------------------------------------------
2022-12-07T19:35:08.7397234Z   format: <class 'dxtbx.format.FormatNXmxDLS.FormatNXmxDLS'>
2022-12-07T19:35:08.7398032Z   num images: 10
2022-12-07T19:35:08.7399119Z   sequences:
2022-12-07T19:35:08.7399874Z     still:    0
2022-12-07T19:35:08.7400537Z     sweep:    1
2022-12-07T19:35:08.7408174Z   num stills: 0
2022-12-07T19:35:08.7409147Z --------------------------------------------------------------------------------
2022-12-07T19:35:08.7424621Z Writing experiments to imported.expt
2022-12-07T19:35:08.7425436Z The following parameters have been modified:
2022-12-07T19:35:08.7426975Z input {
2022-12-07T19:35:08.7427881Z   experiments = imported.expt
2022-12-07T19:35:08.7428700Z }
2022-12-07T19:35:08.7429799Z as_cbf_0001.cbf
2022-12-07T19:35:08.7430833Z Traceback (most recent call last):
2022-12-07T19:35:08.7431678Z   File "%PREFIX%\Scripts\dials.convert_to_cbf-script.py", line 10, in <module>
2022-12-07T19:35:08.7432778Z     sys.exit(run())
2022-12-07T19:35:08.7433494Z   File "%PREFIX%\lib\contextlib.py", line 79, in inner
2022-12-07T19:35:08.7434386Z     return func(*args, **kwds)
2022-12-07T19:35:08.7435827Z   File "%PREFIX%\lib\site-packages\dials\command_line\convert_to_cbf.py", line 77, in run
2022-12-07T19:35:08.7436666Z     convert_to_cbf(imageset, template)
2022-12-07T19:35:08.7437494Z   File "%PREFIX%\lib\site-packages\dials\command_line\convert_to_cbf.py", line 40, in convert_to_cbf
2022-12-07T19:35:08.7438425Z     imageset.get_raw_data(i)[0],
2022-12-07T19:35:08.7439712Z RuntimeError: Please report this error to dials-support@lists.sourceforge.net: dxtbx Error: D:\bld\dxtbx_1670421694546\work\dxtbx\src\dxtbx\..\dxtbx/imageset.h(473): Unknown type long
```

File: src/dxtbx/nexus/__init__.py [hunk 1]
```diff
@@ -487,29 +487,41 @@ def get_static_mask(nxdetector: nxmx.NXdetector) -> tuple[flex.bool, ...] | None
...
         np.half,
         np.single,
-        np.float_,
         np.float16,
-        np.float32,
     )
-    if np.issubdtype(data_np.dtype, np.integer):
-        data_np = data_np.astype(np.int32, copy=False)
-    elif data_np.dtype in np_float_types:
+    if np.issubdtype(data_np.dtype, np.int64):
```
Comment (graeme-winter) [inline src/dxtbx/nexus/__init__.py:500]: Look good

File: src/dxtbx/nexus/__init__.py [hunk 2]
```diff
@@ -487,29 +487,41 @@ def get_static_mask(nxdetector: nxmx.NXdetector) -> tuple[flex.bool, ...] | None
...
+        if bit_depth and bit_depth <= 32:
+            # bit_depth promises that we can safely convert to
+            # a 32-bit int without corrupting the data
+            data_np = data_np.astype(np.intc, copy=False)
+        else:
+            raise TypeError(f"Unsupported dtype {data_np.dtype}")
```
Comment (graeme-winter) [inline src/dxtbx/nexus/__init__.py:506]: 👍

File: src/dxtbx/nexus/__init__.py [hunk 3]
```diff
@@ -498,7 +498,7 @@ def _dataset_as_flex(
         np.float32,
     )
     if np.issubdtype(data_np.dtype, np.integer):
-        data_np = data_np.astype(np.int32, copy=False)
+        data_np = data_np.astype(np.intc, copy=False)
```
[the remark this answers is not in the document]
Comment (graeme-winter) [reply src/dxtbx/nexus/__init__.py:508]: Question above now resolved I think

**Review by graeme-winter — APPROVED** [review]
Looks good to me. Thanks for the additional defences. Appreciate the headaches involved.

**Pull request merged**